File: src/test/java/no/nav/bidrag/dokument/arkiv/JsonMapperTest.java
package no.nav.bidrag.dokument.arkiv;

import static no.nav.bidrag.dokument.arkiv.BidragDokumentArkivConfig.PROFILE_TEST;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import no.nav.bidrag.dokument.arkiv.dto.DokumentoversiktFagsakQuery;
import no.nav.bidrag.dokument.arkiv.dto.EndreJournalpostCommandIntern;
import no.nav.bidrag.dokument.arkiv.dto.Journalpost;
import no.nav.bidrag.dokument.arkiv.dto.OppdaterJournalpostRequest;
import no.nav.bidrag.dokument.dto.EndreDokument;
import no.nav.bidrag.dokument.dto.EndreJournalpostCommand;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.PropertySource;
import org.springframework.test.context.ActiveProfiles;

@ActiveProfiles(PROFILE_TEST)
@DisplayName("Mapping av json verdier")
@PropertySource("classpath:resources.properties")
@SpringBootTest(classes = BidragDokumentArkivLocal.class)
class JsonMapperTest {

  @Autowired
  private ObjectMapper objectMapper;

  @Test
  @DisplayName("skal mappe OppdaterJournalpost til json")
  void skalMappeOppdaterJournalpostTilJson() throws IOException {
    var journalpost = new Journalpost();
    var endreDokument = new EndreDokument();
    endreDokument.setTittel("Tittelen på dokument");
    endreDokument.setDokId(55555);

    var endreJournalpostCommand = new EndreJournalpostCommand();
    endreJournalpostCommand.setAvsenderNavn("AvsenderNavn");
    endreJournalpostCommand.setEndreDokumenter(Arrays.asList(endreDokument));
    endreJournalpostCommand.setFagomrade("BID");
    endreJournalpostCommand.setGjelder("1234");
    endreJournalpostCommand.setTittel("Tittelen på journalposten");
    endreJournalpostCommand.setGjelderType("FNR");
    endreJournalpostCommand.setTilknyttSaker(Arrays.asList("sakIdent"));

    var endreJournalpostIntern = new EndreJournalpostCommandIntern(endreJournalpostCommand, "4805");
    var oppdaterJp = new OppdaterJournalpostRequest(12345, endreJournalpostIntern, journalpost);

    var jsonMap = objectMapper.convertValue(oppdaterJp, Map.class);

    assertAll(
        () -> assertThat(((Map<String, String>)jsonMap.get("avsenderMottaker")).get("navn")).as("avsenderMottaker").isEqualTo("AvsenderNavn"),
        () -> assertThat(((Map<String, String>)jsonMap.get("bruker")).get("id")).as("id").isEqualTo("1234"),
        () -> assertThat(((Map<String, String>)jsonMap.get("bruker")).get("idType")).as("idType").isEqualTo("FNR"),
        () -> assertThat(((Map<String, String>)jsonMap.get("sak")).get("fagsakId")).as("fagsakId").isEqualTo("sakIdent"),
        () -> assertThat(((Map<String, String>)jsonMap.get("sak")).get("fagsaksystem")).as("fagsaksystem").isEqualTo("BISYS"),
        () -> assertThat(((Map<String, String>)jsonMap.get("sak")).get("sakstype")).as("fagsaksystem").isEqualTo("FAGSAK"),
        () -> assertThat(((List<Map<String, String>>)jsonMap.get("dokumenter")).get(0).get("dokumentInfoId")).as("dokumentInfoId").isEqualTo("55555"),
        () -> assertThat(((List<Map<String, String>>)jsonMap.get("dokumenter")).get(0).get("tittel")).as("tittel").isEqualTo("Tittelen på dokument"),
        () -> assertThat(jsonMap.get("tema")).as("tema").isEqualTo("BID"),
        () -> assertThat(jsonMap.get("tittel")).as("tittel").isEqualTo("Tittelen på journalposten")
    );
  }

  @Test
  @DisplayName("skal mappe json streng til java.util.Map")
  void skalMappeJsonRequest() throws JsonProcessingException {
    var opprettJournalpostRequestAsJson = String.join("\n", "{",
        "\"avsenderMottaker\": { \"navn\": \"Birger\" },",
        "\"behandlingstema\": \"BI01\",",
        "\"bruker\": { \"id\": \"06127412345\", \"idType\": \"FNR\" },",
        "\"dokumenter\": [{ \"brevkode\": \"BREVKODEN\", \"dokumentKategori\": \"dokumentKategori\", \"tittel\": \"Tittelen på dokumentet\" }],",
        "\"eksternReferanseId\": \"dokreferanse\",",
        "\"journalfoerendeEnhet\": \"666\",",
        "\"journalpostType\": \"N\",",
        "\"kanal\": \"nav.no\",",
        "\"sak\": { \"arkivsaksnummer\": \"1900001\", \"arkivsaksystem\": \"GSAK\" },",
        "\"tema\": \"BID\",",
        "\"tittel\": \"Tittelen på journalposten\"",
        "}"
    );

    var jsonMap = objectMapper.readValue(opprettJournalpostRequestAsJson, Map.class);

    assertAll(
        () -> assertThat(jsonMap.get("avsenderMottaker")).as("avsenderMottaker").isNotNull(),
        () -> assertThat(jsonMap.get("behandlingstema")).as("behandlingstema").isEqualTo("BI01"),
        () -> assertThat(jsonMap.get("bruker")).as("bruker").isNotNull(),
        () -> assertThat(jsonMap.get("dokumenter")).as("dokumenter").isNotNull(),
        () -> assertThat(jsonMap.get("eksternReferanseId")).as("eksternReferanseId").isEqualTo("dokreferanse"),
        () -> assertThat(jsonMap.get("journalfoerendeEnhet")).as("journalfoerendeEnhet").isEqualTo("666"),
        () -> assertThat(jsonMap.get("journalpostType")).as("journalpostType").isEqualTo("N"),
        () -> assertThat(jsonMap.get("kanal")).as("kanal").isEqualTo("nav.no"),
        () -> assertThat(jsonMap.get("sak")).as("sak").isNotNull(),
        () -> assertThat(jsonMap.get("tema")).as("tema").isEqualTo("BID"),
        () -> assertThat(jsonMap.get("tittel")).as("tittel").isEqualTo("Tittelen på journalposten")
    );
  }

  @Test
  @DisplayName("skal mappe saf query til java.util.Map")
  void skalMappeSafQueryTilMap() throws JsonProcessingException {
    var safQuery = new DokumentoversiktFagsakQuery("666", "BID").getQuery();

    //noinspection unchecked
    assertAll(
        () -> assertThat(safQuery).as("querystring")
            .contains("fagsakId: \"666\"")
            .contains("tema:BID")
    );
  }
}

File: src/test/java/no/nav/bidrag/dokument/arkiv/JsonMapperTest.kt
package no.nav.bidrag.dokument.arkiv

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import no.nav.bidrag.dokument.arkiv.dto.EndreJournalpostCommandIntern
import no.nav.bidrag.dokument.arkiv.dto.JournalStatus
import no.nav.bidrag.dokument.arkiv.dto.Journalpost
import no.nav.bidrag.dokument.arkiv.dto.JournalpostType
import no.nav.bidrag.dokument.arkiv.dto.LagreJournalpostRequest
import no.nav.bidrag.dokument.arkiv.query.DokumentoversiktFagsakQuery
import no.nav.bidrag.dokument.arkiv.query.JournalpostQuery
import no.nav.bidrag.transport.dokument.EndreDokument
import no.nav.bidrag.transport.dokument.EndreJournalpostCommand
import no.nav.bidrag.transport.dokument.IdentType
import no.nav.security.token.support.spring.test.EnableMockOAuth2Server
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles

@ActiveProfiles(BidragDokumentArkivConfig.PROFILE_TEST)
@DisplayName("Mapping av json verdier")
@SpringBootTest(classes = [BidragDokumentArkivTest::class])
@EnableMockOAuth2Server
internal class JsonMapperTest {
    @Autowired
    private lateinit var objectMapper: ObjectMapper

    @Test
    @DisplayName("skal mappe OppdaterJournalpost til json")
    fun skalMappeOppdaterJournalpostTilJson() {
        val journalpost = Journalpost()
        journalpost.journalstatus = JournalStatus.MOTTATT
        journalpost.journalposttype = JournalpostType.I
        val endreDokument = EndreDokument(null, "55555", "55555", "Tittelen på dokument")
        val endreJournalpostCommand = EndreJournalpostCommand(
            avsenderNavn = "AvsenderNavn",
            endreDokumenter = listOf(endreDokument),
            fagomrade = "BID",
            gjelder = "1234",
            tittel = "Tittelen på journalposten",
            gjelderType = IdentType.FNR,
            tilknyttSaker = listOf("sakIdent")
        )
        val endreJournalpostIntern = EndreJournalpostCommandIntern(endreJournalpostCommand, "4805")
        val oppdaterJp = LagreJournalpostRequest(12345, endreJournalpostIntern, journalpost)
        val jsonMap = objectMapper.convertValue(
            oppdaterJp,
            MutableMap::class.java
        ) as Map<String, Any>
        val jsonObjects = JsonObjects(jsonMap)
        org.junit.jupiter.api.Assertions.assertAll(
            Executable {
                Assertions.assertThat(
                    jsonObjects.objekt("avsenderMottaker")!!["navn"]
                )
                    .`as`("avsenderMottaker")
                    .isEqualTo("AvsenderNavn")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.objekt("bruker")!!["id"]
                ).`as`("id").isEqualTo("1234")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.objekt("bruker")!!["idType"]
                ).`as`("idType").isEqualTo("FNR")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.objekt("sak")!!["fagsakId"]
                )
                    .`as`("fagsakId")
                    .isEqualTo("sakIdent")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.objekt("sak")!!["fagsaksystem"]
                )
                    .`as`("fagsaksystem")
                    .isEqualTo("BISYS")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.objekt("sak")!!["sakstype"]
                )
                    .`as`("fagsaksystem")
                    .isEqualTo("FAGSAK")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.listeMedObjekter("dokumenter")!![0]["dokumentInfoId"]
                )
                    .`as`("dokumentInfoId")
                    .isEqualTo("55555")
            },
            Executable {
                Assertions.assertThat(
                    jsonObjects.listeMedObjekter("dokumenter")!![0]["tittel"]
                )
                    .`as`("tittel")
                    .isEqualTo("Tittelen på dokument")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["tema"]
                ).`as`("tema").isEqualTo("BID")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["tittel"]
                ).`as`("tittel").isEqualTo("Tittelen på journalposten")
            }
        )
    }

    @Test
    @DisplayName("skal mappe json streng til java.util.Map")
    @Throws(
        JsonProcessingException::class
    )
    fun skalMappeJsonRequest() {
        val opprettJournalpostRequestAsJson = java.lang.String.join(
            "\n",
            "{",
            "\"avsenderMottaker\": { \"navn\": \"Birger\" },",
            "\"behandlingstema\": \"BI01\",",
            "\"bruker\": { \"id\": \"06127412345\", \"idType\": \"FNR\" },",
            "\"dokumenter\": [{ \"brevkode\": \"BREVKODEN\", \"dokumentKategori\": \"dokumentKategori\", \"tittel\": \"Tittelen på dokumentet\" }],",
            "\"eksternReferanseId\": \"dokreferanse\",",
            "\"journalfoerendeEnhet\": \"666\",",
            "\"journalpostType\": \"N\",",
            "\"kanal\": \"nav.no\",",
            "\"sak\": { \"arkivsaksnummer\": \"1900001\", \"arkivsaksystem\": \"GSAK\" },",
            "\"tema\": \"BID\",",
            "\"tittel\": \"Tittelen på journalposten\"",
            "}"
        )
        val jsonMap: Map<*, *> = objectMapper.readValue(
            opprettJournalpostRequestAsJson,
            MutableMap::class.java
        )
        org.junit.jupiter.api.Assertions.assertAll(
            Executable {
                Assertions.assertThat(
                    jsonMap["avsenderMottaker"]
                ).`as`("avsenderMottaker").isNotNull()
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["behandlingstema"]
                ).`as`("behandlingstema").isEqualTo("BI01")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["bruker"]
                ).`as`("bruker").isNotNull()
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["dokumenter"]
                ).`as`("dokumenter").isNotNull()
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["eksternReferanseId"]
                )
                    .`as`("eksternReferanseId")
                    .isEqualTo("dokreferanse")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["journalfoerendeEnhet"]
                )
                    .`as`("journalfoerendeEnhet")
                    .isEqualTo("666")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["journalpostType"]
                ).`as`("journalpostType").isEqualTo("N")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["kanal"]
                ).`as`("kanal").isEqualTo("nav.no")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["sak"]
                ).`as`("sak").isNotNull()
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["tema"]
                ).`as`("tema").isEqualTo("BID")
            },
            Executable {
                Assertions.assertThat(
                    jsonMap["tittel"]
                ).`as`("tittel").isEqualTo("Tittelen på journalposten")
            }
        )
    }

    @Test
    @DisplayName("skal mappe saf dokumentOversiktFagsak query til java.util.Map")
    fun skalMappeSafDokumentOversiktQueryTilMap() {
        val safQuery = DokumentoversiktFagsakQuery("666", listOf("BID"))
        org.junit.jupiter.api.Assertions.assertAll(
            Executable {
                Assertions.assertThat(safQuery.getQuery())
                    .`as`("querystring")
                    .contains("fagsakId: \$fagsakId")
                    .contains("tema:\$tema")
            },
            Executable {
                Assertions.assertThat(safQuery.getVariables())
                    .`as`("Variables")
                    .containsEntry("fagsakId", "666")
                    .containsEntry("tema", listOf("BID"))
            }
        )
    }

    @Test
    @DisplayName("skal mappe saf journalpost query til java.util.Map")
    fun skalMappeSafJournalpostQueryTilMap() {
        val safQuery = JournalpostQuery(1235L)
        org.junit.jupiter.api.Assertions.assertAll(
            Executable {
                Assertions.assertThat(safQuery.getQuery())
                    .`as`("querystring")
                    .contains("journalpostId: \$journalpostId")
            },
            Executable {
                Assertions.assertThat(safQuery.getVariables())
                    .`as`("Variables")
                    .containsEntry("journalpostId", "1235")
            }
        )
    }

    private class JsonObjects(private val jsonMap: Map<String, Any>) {
        fun objekt(key: String): Map<String, String>? {
            return jsonMap[key] as Map<String, String>?
        }

        fun listeMedObjekter(key: String): List<Map<String, Any>>? {
            return jsonMap[key] as List<Map<String, Any>>?
        }
    }
}
